sources/host-ctr/cmd/host-ctr/main.go

```diff
@@ -279,30 +280,20 @@ func runCtr(containerdSocket string, namespace string, containerID string, sourc
 			oci.WithHostNamespace(runtimespec.NetworkNamespace),
 			oci.WithHostHostsFile,
 			oci.WithHostResolvconf,
-			// Mount in the API socket for the Bottlerocket API server, and the API
-			// client used to interact with it
-			oci.WithMounts([]runtimespec.Mount{
-				{
-					Options:     []string{"bind", "rw"},
-					Destination: "/run/api.sock",
-					Source:      "/run/api.sock",
-				},
-				// Mount in the apiclient to make API calls to the Bottlerocket API server
-				{
-					Options:     []string{"bind", "ro"},
-					Destination: "/usr/local/bin/apiclient",
-					Source:      "/usr/bin/apiclient",
-				}}),
 			// Pass proxy environment variables to this container
 			withProxyEnv(),
-			// Mount in the persistent storage location for this container
-			withPersistentStorage(containerName, persistentDir),
-			// Mount the rootfs with an SELinux label that makes it writable
+			// Mount all the required endpoints for this container
+			withMounts(privileged, containerName, persistentDir),
+			// Mount the container's rootfs with an SELinux label that makes it writable
 			withMountLabel("system_u:object_r:secret_t:s0"),
 			// Include conditional options for superpowered containers.
 			withSuperpowered(superpowered),
-			// Mount the rootfs if superpowered or bootstrap
-			withRootFilesystemMounts(superpowered || cType == bootstrap),
+			// Mount privileged containers's rootfs with shared propagations
+			withRootFsShared(privileged),
+			// Set the privileges for this container
+			withPrivileges(superpowered, cType == bootstrap),
+			// Set what devices are allowed for this container
+			withDevices(privileged),
```

Comment:
I think this is starting to get unwieldy and the logic for "what options are set on the spec" is mixed between this function and all these helper functional options.

I think a somewhat better structure would separate the two and make the code easier to reason about:

```go
specOpts := []oci.SpecOpts{
	// basic spec opts shared between all containers
	oci.WithImageConfig(img),
	oci.WithHostNamespace(runtimespec.NetworkNamespace),
	oci.WithHostHostsFile,
	oci.WithHostResolvconf,
	withProxyEnv(),
	withDefaultMounts(),
	withMountLabel("system_u:object_r:secret_t:s0"),
}
switch {
case superpowered:
	specOpts = append(specOpts, withSuperpowered())
case cType == bootstrap:
	specOpts = append(specOpts, withBootstrap())
default:
	specOpts = append(specOpts, withDefault())
}
``` 

```go
func withSuperpowered() oci.SpecOpts {
	return oci.Compose(
		oci.WithSelinuxLable(...),
		withPrivilegedMounts(),
		withRootFsShared(),
		withPrivileged(),
		withDevices(),
}

func withBootstrap() oci.SpecOpts {
	return oci.Compose(
		withPrivilegedMounts(),
		withRootFsShared(),
		withCapSysAdmin(),
		withDevices(),
	)
}

func withDefault() oci.SpecOpts {
	return oci.Compose(
		// if there's anything that even has to go here...
	)
}
```